docs/designs/evaluation_design.md

```diff
@@ -0,0 +1,63 @@
+As shown in <em>Figure 1</em>. The worker 1 gets model from the pserver at the 15th step and the worker 2 gets model later at 25th step for the reason that the worker 2 is slower. However, the pserver has updated the model by gradients from the worker 2 before the worker2 gets the model to evaluate. So records assigned to different worker in the same validation dataset will be evaluated with different model version. 
+
+
+## Evaluate with Current Model in Pserver 
+1. The pserver updates model variables if it receives gradients from the worker after an evaluation job starts.\
+The solution is the same as existing design in ElasticDL which can not resolve Problem 2. To resolve Problem 1, the master can divide train and evaluation tasks into two todo lists at the 10th step in <em>Figure 1</em>. When evaluation starts, the master generates evaluation tasks and inserts those into the evaluation todo list not the train list. Meanwhile, the master needs to tell all pservers to change the model mode to evaluation.  The worker starts to pull evaluation tasks and inferences outputs when the mode of model it gets is evaluation. The pserver updates model variables by the mini-batch gradients from each worker at most once. Because the model mode is evaluating when the worker gets for next mini-batch. The master will change the model mode to train after receiving all record outputs of evaluation tasks. Then worker(s) continues to process training tasks.
+
+![evaluate_flowchart_proposal](/docs/images/train_and_evaluate_flowchart_proposal.svg)\
+<em>Figure 3</em>. Proposal Flowchart of Train and Evaluation
+
+2. Pserver stops to update model variables when an evaluation job starts.\
+As shown in <em>Figure 1</em>, pserver will not execute the 18th and 19th steps after an evaluation job starts at the 10th step. So, the model worker 2 get is the same as the worker 1. The solution can resolve Problem 2 and also can resolve Problem 1 by solution in <em>Figure 3.</em>. But, the training process must wait all evaluation tasks have been completed. 
+
+Questions to discuss:
+* None of training checkpoints will be evaluated in above two solutions which are different from tf.estimator. tf.estimator will load the last checkpoint saved by train to evaluate in a separated pod named evaluator. 
+* Should master preserve the model to checkpoint after evaluation?
+
+## Evaluate by Loading Checkpoint to Pserver
+Generally, deep learning frameworks preserve model as checkpoint every many seconds or steps. So pserver can load the model from checkpoint for evaluation. There are also two designs for pserver to load model. 
+
+1. The master launches additional pods as evaluation pserver.\
+When an evaluation job starts, the master will launch additional pods as evaluation pservers and evaluation pservers will load the model from checkpoint.\
```

Comment:
How many ps pods should we launch? The same as the training ps?

Reply:
The same number as the training ps may be a choice. If the model in checkpoint can be loaded by the different number of ps pods, the least number of pods may be launched.